fix(anthropic): reorder tool_result parts to front of combined user messages

tayyab3245 opened the pull request:
Fixes #8318
## Description


fix(anthropic): reorder tool_result parts to front of combined user messages
- Addresses Claude API validation requiring tool_result immediately after tool_use
- Preserves role alternation by reordering within combined user messages
- Maintains message combining behavior from #2067 that solved #2047
- Added comprehensive tests for single/multi-tool scenarios
- **Verified with live Anthropic Claude API** - tested actual tool calling scenarios
- All existing tests continue to pass (101/101)




---

## Background

### Why was this change necessary?

Users were encountering a Claude API error when using tool calling functionality:

```
"tool_use ids were found without tool_result blocks immediately after: tool-example-123. 
Each tool_use block must have a corresponding tool_result block in the next message."
```

This error occurred because the AI SDK's Anthropic message converter was placing user text before `tool_result` content in combined user messages, violating Claude's strict validation requirements.


**The Problem:**
- Assistant sends `tool_use` → expects `tool_result` in the very next message
- AI SDK was combining: `[tool_result, user_text]` in one message
- Claude API rejected this as malformed

## Summary

**What did you change?**

Modified the user message processing in `convert-to-anthropic-messages-prompt.ts`:

1. **Added content reordering logic**: Filter and separate `tool_result` vs other content parts
2. **Preserved message combining**: Maintains the role alternation fix from #2067 that solved #2047
3. **Ensured Claude compliance**: `tool_result` parts now appear first in combined messages
4. **Updated tests**: Added 3 new comprehensive tests covering single/multi-tool scenarios

### Result

- **Before**: `[{role: 'user', content: [user_text, tool_result]}]`
- **After**: `[{role: 'user', content: [tool_result, user_text]}]`

## Manual Verification

**How I manually verified the fix:**

1. **Reproduced the original error**: Used the exact message sequence from #8318
2. **Confirmed fix resolves error**: Same sequence now works without Claude API rejection  
3. **Live API testing with real Claude endpoints**: Created actual tool calling scenarios and verified the fix works with Anthropic's production API
4. **Tested edge cases**: Multiple tool results, error tool results, content ordering
5. **Verified backward compatibility**: All existing tests pass (101/101)
6. **Preserved role alternation**: Confirmed no regression of the #2047 fix

**Test Scenarios Verified:**

- Single tool call followed by user message (the main bug)
- Multiple tool results with user text in same message
- Error tool results (`is_error: true`) with user content
- Content part ordering preservation within reordered messages

## Tasks

**PR Checklist:**

- [x] **Tests have been added / updated** (for bug fixes / features)
  - Updated existing test: "should combine tool and user messages with tool_result first"
  - Added new tests: 3 comprehensive scenarios covering reordering logic
  - All 101 tests in @ai-sdk/anthropic package pass
  
- [x] **Documentation has been added / updated** (for bug fixes / features)
  - Added clear comments explaining Claude's validation requirements
  - Documented preservation of role alternation behavior from #2067
  
- [x] **A patch changeset for relevant packages has been added** (for bug fixes / features)
  - Created changeset: `pretty-carpets-cheer.md` for `@ai-sdk/anthropic`
  
- [x] **Formatting issues have been fixed** (run `pnpm prettier-fix`)
  - Code follows consistent formatting patterns, all Prettier checks pass
  
- [x] **I have reviewed this pull request** (self-review)
  - Verified minimal scope, surgical fix preserving existing behavior
  - Confirmed all edge cases handled with comprehensive test coverage



## Future Work

**Potential follow-up improvements (not in scope for this bug fix):**

1. **Performance optimization**: Cache content filtering results for large message arrays
2. **Enhanced testing**: Add performance benchmarks for message conversion with tools
3. **Error messaging**: Provide more descriptive errors when Claude validation fails
4. **Documentation**: Add examples to docs showing proper tool calling message flows

*Note: These are enhancements, not requirements for fixing the reported bug.*

## Related Issues

- **Fixes #8318** - generateObject fails using Claude provider with "tool_use ids were found without tool_result" error
- **Preserves #2067** - Message combining behavior that solved role alternation issue #2047
- **Addresses**: Claude API message sequencing requirements for tool calling


---

## Technical Details

**Files Changed:**
- `packages/anthropic/src/convert-to-anthropic-messages-prompt.ts` (core reordering logic)
- `packages/anthropic/src/convert-to-anthropic-messages-prompt.test.ts` (comprehensive test updates)
- `.changeset/pretty-carpets-cheer.md` (release notes)

### Key Implementation

- `packages/anthropic/src/convert-to-anthropic-messages-prompt.ts` (core reordering logic)
- `packages/anthropic/src/convert-to-anthropic-messages-prompt.test.ts` (comprehensive test updates)
- `.changeset/pretty-carpets-cheer.md` (release notes)

tayyab3245 committed:
fix(anthropic): separate tool_result from user content to fix Claude API error

- Modified groupIntoBlocks to ensure tool messages create separate user blocks
- Prevents Claude API error: 'tool_use ids were found without tool_result blocks immediately after'
- Added regression test reproducing the exact scenario from issue #8318
- All existing tests continue to pass (98/98)

Fixes #8318

gr2m committed:
Merge branch 'main' into fix/claude-tool-result-message-separation

gr2m commented on packages/anthropic/src/convert-to-anthropic-messages-prompt.test.ts:
```diff
@@ -372,7 +372,7 @@ describe('tool messages', () => {
   });
 
-  it('should combine user and tool messages', async () => {
+  it('should separate tool and user messages', async () => {
```
I'm hesitant to just change the current behavior. Given that it has a test, it might have been intentional. 

@lgrammel introduced this test as part of https://github.com/vercel/ai/pull/2067, which in turn fixed https://github.com/vercel/ai/issues/2047

> AI_APICallError: messages: roles must alternate between "user" and "assistant", but found multiple "user" roles in a row.

I think this "fix" could re-introduce the above error message? Could you look into that and make sure we don't introduce a regression?

tayyab3245 replied:
@gr2m Thanks for the thoughtful feedback and for pointing out the potential for a regression. I'll dig into the history of the related issues you mentioned and will report back here with my findings.

tayyab3245 committed:
fix: reorder tool_result parts first in combined messages

- Addresses Claude API validation requiring tool_result immediately after tool_use
- Preserves role alternation by reordering within combined user messages
- Maintains message combining behavior from #2067 that solved #2047
- Added comprehensive tests for single/multi-tool scenarios

gr2m commented on packages/anthropic/src/convert-to-anthropic-messages-prompt.ts:
```diff
@@ -554,6 +568,7 @@ function groupIntoBlocks(
   const blocks: Array<SystemBlock | AssistantBlock | UserBlock> = [];
   let currentBlock: SystemBlock | AssistantBlock | UserBlock | undefined =
     undefined;
+  let lastWasToolMessage = false;
```
This one is correct, just remove `lastWasToolMessage`

gr2m committed:
remove unused variable

resolves https://github.com/vercel/ai/pull/8474#discussion_r2328493158

gr2m reviewed (commented):
can you update the pull request title as well? That will be used in the release notes

tayyab3245 committed:
Address maintainer feedback: update changeset description and test names

- Update changeset to reflect reordering approach instead of separation
- Remove issue number from test name per maintainer guidance
- Make test descriptions focus on actual behavior (tool_result before text)

gr2m reviewed (approved):
Great first PR to the AI SDK! Congratulations! And thank you 💐

lgrammel commented:
Reverting because of #8516
